File: LeetCode/1649. Create Sorted Array through Instructions/solution.py
class Solution:
    def createSortedArray(self, instructions: List[int]) -> int:
        n = max(instructions)
        arr = [0]*(n+1)
        def query(x):
            cnt = 0
            while x > 0:
                cnt += arr[x]
                x -= (x&(-x))
            return cnt
        def update(x):
            while x <= n:
                arr[x] += 1
                x += (x&(-x))
            return
        res, MOD = 0, 10**9+7
        for i, cur in enumerate(instructions):
            res += min(query(cur-1),i-query(cur))
            res %= MOD
            update(cur)
        return res 

File: LeetCode/1649. Create Sorted Array through Instructions/solution.cpp
using ll = long long;

const int MOD = 1e9+7;
const int maxn = 1e5+10;
int tree[maxn];

class Solution {
public:
    void insert(int x){
        while (x <= maxn){
            tree[x]++;
            x += (x&(-x));
        }
    }
    int query(int x){
        int sums = 0;
        while (x){
            sums += tree[x];
            x -= (x&(-x));
        }
        return sums;
    }
    int createSortedArray(vector<int>& instructions) {
        memset(tree, 0, maxn*sizeof(int));
        ll res = 0;
        int cnt = 0;
        for (auto x : instructions){
            res += min(query(x-1), cnt-query(x));
            res %= MOD;
            cnt++;
            insert(x);
        }
        return res;

    }
};